src/affine_dimensions.jl

```diff
@@ -0,0 +1,466 @@
+function AffineDimensions{R}(s::Real, o::UnionAbstractQuantity, q::UnionAbstractQuantity, sym::Symbol=:nothing) where R
+    q0   = siunits(0*q) #Origin point in SI units
+    oΔ   = siunits(o) - siunits(0*o) #Offset is a difference in affine units
+    dimension(q0) == dimension(oΔ) || throw(DimensionError(o, q)) #Check the units and give an informative error
+    
+    #Obtain SI units of the scale and offset
+    o_si = oΔ + q0 #Total offset is origin plus the offset
+    q_si = siunits(q) - q0 #The scaling quantity must remove the origin
+    
+    #Call the SI quantity constructor
+    return AffineDimensions{R}(s, o_si, q_si, sym)
+end
+
+#Base case when everyting is convrted to si units (offset is assumed to be in SI units)
+function AffineDimensions{R}(s::Real, o::UnionAbstractQuantity{<:Any,<:Dimensions}, q::UnionAbstractQuantity{<:Any,<:Dimensions}, sym::Symbol=:nothing) where R
+    dimension(o) == dimension(q) || throw(DimensionError(o, q))
+    o_val = ustrip(o)
+    q_val = ustrip(q)
+    return AffineDimensions{R}(s*q_val, o_val, dimension(q), sym)
+end
+
+#If a quantity is used only for the dimension, the offset is assumed to be in the same scale as the quantity 
+function AffineDimensions{R}(s::Real, o::Real, q::Q, sym::Symbol=:nothing) where {R, Q<:UnionAbstractQuantity}
+    return AffineDimensions{R}(s, o*q, q, sym)
+end
+
+
+scale(d::AffineDimensions)  = d.scale
+offset(d::AffineDimensions) = d.offset
+basedim(d::AffineDimensions) = d.basedim
+
+with_type_parameters(::Type{<:AffineDimensions}, ::Type{R}) where {R} = AffineDimensions{R}
+constructorof(::Type{AffineDimensions}) = AffineDimensions{DEFAULT_DIM_BASE_TYPE}
+constructorof(::Type{AffineDimensions{R}}) where R = AffineDimensions{R}
+
+function Base.show(io::IO, d::AbstractAffineDimensions)
+    addsign = ifelse(offset(d)<0, "-" , "+")
+
+    if d.symbol != :nothing
+        print(io, d.symbol)
+    elseif isone(scale(d)) & iszero(offset(d))
+        print(io, basedim(d))
+    elseif iszero(offset(d))
+        print(io, "(", scale(d), " ", basedim(d),")")
+    elseif isone(scale(d))
+        print(io, "(", addsign, abs(offset(d)), basedim(d), ")")
+    else
+        print(io, "(", scale(d), addsign, abs(offset(d)), " ", basedim(d),")")
+    end
+end
+
+assert_no_offset(d::AffineDimensions) = iszero(offset(d)) || throw(AssertionError("AffineDimensions $(d) has a non-zero offset, implicit conversion is not allowed due to ambiguity. Use uexpand(x) to explicitly convert"))
+siunits(q::UnionAbstractQuantity{<:Any, <:Dimensions}) = q
+siunits(q::UnionAbstractQuantity{<:Any, <:AbstractSymbolicDimensions}) = uexpand(q)
+function siunits(q::Q) where {T, R, D<:AbstractAffineDimensions{R}, Q<:UnionAbstractQuantity{T,D}}
+    return force_convert(with_type_parameters(Q, T, Dimensions{R}), q)
+end
+siunits(q::QuantityArray) = siunits.(q)
+
+
+"""
+uexpand(q::Q) where {T, R, D<:AbstractAffineDimensions{R}, Q<:UnionAbstractQuantity{T,D}}
```

Comment:
Please format docstrings with 4 leading spaces preceding the signature, similar to the other docstrings